--- charts-original/templates/_helpers.tpl
+++ charts/templates/_helpers.tpl
@@ -7,7 +7,11 @@
 For backwards compatibility, if fullnameOverride is not kiali but deployment.instance_name is kiali,
 use fullnameOverride, otherwise use deployment.instance_name.
 */}}
-{{- define "kiali-server.fullname" -}}
+{{- if (and .Release.IsUpgrade $deployment)}}
+app.kubernetes.io/instance: {{ (get (($deployment).metadata.labels) "app.kubernetes.io/instance") | default $fullName }}
+{{- else }}
+app.kubernetes.io/instance: {{ $fullName }}
+{{- end }}
 {{- if (and (eq .Values.deployment.instance_name "kiali") (ne .Values.fullnameOverride "kiali")) }}
   {{- .Values.fullnameOverride | trunc 63 }}
 {{- else }}
@@ -40,6 +44,9 @@
 helm.sh/chart: {{ include "kiali-server.chart" . }}
 app: kiali
 {{ include "kiali-server.selectorLabels" . }}
+{{- $releaseName := .Release.Name -}}
+{{- $fullName := include "kiali-server.fullname" . -}}
+{{- $deployment := (lookup "apps/v1" "Deployment" .Release.Namespace $fullName) -}}
 version: {{ .Values.deployment.version_label | default .Chart.AppVersion | quote }}
 app.kubernetes.io/version: {{ .Values.deployment.version_label | default .Chart.AppVersion | quote }}
 app.kubernetes.io/managed-by: {{ .Release.Service }}
@@ -170,4 +177,27 @@
 {{- else }}
   {{- .Release.Namespace }}
 {{- end }}
-{{- end }}
\ No newline at end of file
+{{- end }}
+
+{{- define "system_default_registry" -}}
+{{- if .Values.global.cattle.systemDefaultRegistry -}}
+{{- printf "%s/" .Values.global.cattle.systemDefaultRegistry -}}
+{{- else -}}
+{{- "" -}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Windows cluster will add default taint for linux nodes,
+add below linux tolerations to workloads could be scheduled to those linux nodes
+*/}}
+{{- define "linux-node-tolerations" -}}
+- key: "cattle.io/os"
+  value: "linux"
+  effect: "NoSchedule"
+  operator: "Equal"
+{{- end -}}
+
+{{- define "linux-node-selector" -}}
+kubernetes.io/os: linux
+{{- end -}}
